### PR TITLE
synchronized ADS event handling in WorkspaceSelector

### LINKED ISSUE #32032: Investigate hard crash in workbench
Workbench crashes sporadically, especially when running a script that generates a lot of workspaces and groups in ADS.
Most likely, it's a regression introduced recently (such as the workspace selectors in the new workspace calculator).

### Expected behavior
No hard crash.

### Actual behavior
Crashes frequently, but sporadically.

### Steps to reproduce the behavior
```
from mantid.simpleapi import *
ws_list = []

for i in range(50000):
    print(i)
    name = "ws" + str(i)
    CreateSingleValuedWorkspace(OutputWorkspace=name, DataValue=0)
    ws_list.append(name)
    if i % 10 == 0:     
        DeleteWorkspaces(WorkspaceList=ws_list)
        ws_list.clear()
```

### Platforms affected
All.

### Crash trace
```
Python(24955,0x10a02be00) malloc: *** error for object 0x7fc39bdf52c0: pointer being freed was not allocated Python(24955,0x10a02be00) malloc: *** set a breakpoint in malloc_error_break to debug

This is the full trace:

Process:               Python [25016]
Path:                  /usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/Resources/Python.app/Contents/MacOS/Python
Identifier:            org.python.python
Version:               3.9.5 (3.9.5)
Code Type:             X86-64 (Native)
Parent Process:        MantidWorkbench [25015]
Responsible:           Terminal [2500]
User ID:               103532

Date/Time:             2021-07-12 12:10:17.461 +0200
OS Version:            macOS 11.4 (20F71)
Report Version:        12
Anonymous UUID:        E55A4C87-6D73-E62D-EDF8-D8ABCABA6D5F

Sleep/Wake UUID:       83CC6A2E-361E-46B2-955A-926DD06AE793

Time Awake Since Boot: 210000 seconds
Time Since Wake:       3700 seconds

System Integrity Protection: disabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       EXC_I386_GPFLT
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [25016]

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   org.qt-project.QtGui          	0x0000000133521327 QStandardItemModel::data(QModelIndex const&, int) const + 119
1   org.qt-project.QtWidgets      	0x000000013405a037 0x133cc3000 + 3764279
2   org.qt-project.QtWidgets      	0x000000013405a768 0x133cc3000 + 3766120
3   org.qt-project.QtWidgets      	0x0000000134057a7d 0x133cc3000 + 3754621
4   org.qt-project.QtWidgets      	0x000000013405754a 0x133cc3000 + 3753290
5   org.qt-project.QtCore         	0x0000000128987614 QObject::event(QEvent*) + 900
6   org.qt-project.QtWidgets      	0x0000000133cd6a56 QApplicationPrivate::notify_helper(QObject*, QEvent*) + 262
7   org.qt-project.QtWidgets      	0x0000000133cd7e02 QApplication::notify(QObject*, QEvent*) + 466
8   QtWidgets.abi3.so             	0x000000013397319a sipQApplication::notify(QObject*, QEvent*) + 122
9   org.qt-project.QtCore         	0x000000012895e2b6 QCoreApplication::notifyInternal2(QObject*, QEvent*) + 166
10  org.qt-project.QtCore         	0x000000012895f3f3 QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) + 803
11  libqcocoa.dylib               	0x000000014206e892 0x142032000 + 247954
12  libqcocoa.dylib               	0x000000014206ef98 0x142032000 + 249752
13  com.apple.CoreFoundation      	0x00007fff2041e6dc __CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE0_PERFORM_FUNCTION__ + 17
14  com.apple.CoreFoundation      	0x00007fff2041e644 __CFRunLoopDoSource0 + 180
15  com.apple.CoreFoundation      	0x00007fff2041e3ba __CFRunLoopDoSources0 + 242
16  com.apple.CoreFoundation      	0x00007fff2041cdc8 __CFRunLoopRun + 897
17  com.apple.CoreFoundation      	0x00007fff2041c380 CFRunLoopRunSpecific + 567
18  com.apple.HIToolbox           	0x00007fff288bcab3 RunCurrentEventLoopInMode + 292
19  com.apple.HIToolbox           	0x00007fff288bc815 ReceiveNextEventCommon + 587
20  com.apple.HIToolbox           	0x00007fff288bc5b3 _BlockUntilNextEventMatchingListInModeWithFilter + 70
21  com.apple.AppKit              	0x00007fff22b476f2 _DPSNextEvent + 864
22  com.apple.AppKit              	0x00007fff22b45ec5 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 1364
23  com.apple.AppKit              	0x00007fff22b38239 -[NSApplication run] + 586
24  libqcocoa.dylib               	0x000000014206dd74 0x142032000 + 245108
25  org.qt-project.QtCore         	0x000000012895a6b7 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 471
26  org.qt-project.QtCore         	0x000000012895e852 QCoreApplication::exec() + 130
27  QtWidgets.abi3.so             	0x00000001339740d0 meth_QApplication_exec_(_object*, _object*) + 80
28  org.python.python             	0x000000010bb970c2 cfunction_call + 90
29  org.python.python             	0x000000010bb4c6da _PyObject_MakeTpCall + 129
30  org.python.python             	0x000000010bc44cbd call_function + 270
31  org.python.python             	0x000000010bc42526 _PyEval_EvalFrameDefault + 45324
32  org.python.python             	0x000000010bb4d240 _PyFunction_Vectorcall + 192
33  org.python.python             	0x000000010bc44c53 call_function + 164
34  org.python.python             	0x000000010bc42562 _PyEval_EvalFrameDefault + 45384
35  org.python.python             	0x000000010bb4d240 _PyFunction_Vectorcall + 192
36  org.python.python             	0x000000010bc44c53 call_function + 164
37  org.python.python             	0x000000010bc42562 _PyEval_EvalFrameDefault + 45384
38  org.python.python             	0x000000010bb4d240 _PyFunction_Vectorcall + 192
39  org.python.python             	0x000000010bc44c53 call_function + 164
40  org.python.python             	0x000000010bc42562 _PyEval_EvalFrameDefault + 45384
41  org.python.python             	0x000000010bb4d240 _PyFunction_Vectorcall + 192
42  org.python.python             	0x000000010bc44c53 call_function + 164
43  org.python.python             	0x000000010bc42562 _PyEval_EvalFrameDefault + 45384
44  org.python.python             	0x000000010bc362d6 _PyEval_EvalCode + 403
45  org.python.python             	0x000000010bc3104f builtin_exec + 389
46  org.python.python             	0x000000010bb97a43 cfunction_vectorcall_FASTCALL + 95
47  org.python.python             	0x000000010bc44c53 call_function + 164
48  org.python.python             	0x000000010bc42562 _PyEval_EvalFrameDefault + 45384
49  org.python.python             	0x000000010bc362d6 _PyEval_EvalCode + 403
50  org.python.python             	0x000000010bb4d2f8 _PyFunction_Vectorcall + 376
51  org.python.python             	0x000000010bc44c53 call_function + 164
52  org.python.python             	0x000000010bc42562 _PyEval_EvalFrameDefault + 45384
53  org.python.python             	0x000000010bc362d6 _PyEval_EvalCode + 403
54  org.python.python             	0x000000010bb4d2f8 _PyFunction_Vectorcall + 376
55  org.python.python             	0x000000010bcb4fd0 pymain_run_module + 212
56  org.python.python             	0x000000010bcb4a9a pymain_run_python + 433
57  org.python.python             	0x000000010bcb48ad Py_RunMain + 23
58  org.python.python             	0x000000010bcb5bca pymain_main + 35
59  org.python.python             	0x000000010bcb5ea0 Py_BytesMain + 42
60  libdyld.dylib                 	0x00007fff20340f5d start + 1
```

### PR DESCRIPTION
**Pull request by darominski**
**Description of work.**

It has been found that relatively rapid changes in the ADS contents can cause hard crash coming from the `WorkspaceSelector`. The issue is revealed as the new `WorkspaceCalculator` widget has 3 selectors and is docked, so it is present all the time. The main issue was that the `WorkspaceSelector` observes the ADS and the handlers were not thread protected. If a notification comes while handling another one, the structure of the underlying QComboBox is not consistent. 
 
The proposed solution is as follows:
1. Make ADS event handlers in the `WorkspaceSelector` synchronized.
2. In `WorkspaceCalculator` :
    a) Do not observe ADS if the widget is hidden, this is for extra security.
    b) `WorkspaceSelector`s are set to not observe ADS at the construction time
    c) They are set to observe on`focusInEvent` 
    d) They disconnect on `activated` signal

There are still two scenarios to cause a crash, which will be investigated separately.
This is when one opens a selector but doesn't make a choice and keeps it open. This means that focusIn is triggered, but activate is not, so the widget stays in observing state. The other is when a selector is opened while the script is running.
However, this PR already reduces the chance of the crash to a minimum, so it should get in regardless.

**To test:**
The following script exercises the issue:
Note that it creates numerous small (single-valued) workspaces in a script; a scenario that induces the crash.
With real-sized workspaces this doesn't occur as the creation/manipulation time of the workspaces gives enough time for the widget to react.

```
from mantid.simpleapi import *
ws_list = []

for i in range(50000):
    print(i)
    name = "ws" + str(i)
    CreateSingleValuedWorkspace(OutputWorkspace=name, DataValue=0)
    ws_list.append(name)
    if i % 10 == 0:     
        DeleteWorkspaces(WorkspaceList=ws_list)
        ws_list.clear()
```

Case 1:
1. Hide WorkspaceCalculator using the menu
2. Run the script, make sure there is no crash

Case 2:
1. Show the WorkspaceCalculator, but don't touch it
2. Run the script, make sure there is no crash

Case 3:
1. Run the script, and interrupt at some point
2. Click around on the widget, choose some workspaces in the selectors
3. Run the script, make sure there is no crash

Case 4 (still crashing):
1. Click on one of the selectors, but do not choose anything
2. Run the script while the list is open (you will need to do Ctrl+Enter to avoid clicking)

Case 5 (still crashing):
1. Run the script
2. Click on one of the selectors while the script is running

Ideas are welcome for 4 and 5, but this is a type of problem that is hard to track down, since the crash doesn't happen on debug build. 

Fixes #32032 . 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gvardany** [comment]
@darominski I think there is some misunderstanding. What I meant is that we should add a switch to the `WorkspaceSelector` whether to observe the ADS. The host GUI should decide if that makes sense or not. We should not completely remove that, as it is used in many places and might be expected to work as is elsewhere.

**Comment by darominski** [comment]
> @darominski I think there is some misunderstanding. What I meant is that we should add a switch to the `WorkspaceSelector` whether to observe the ADS. The host GUI should decide if that makes sense or not. We should not completely remove that, as it is used in many places and might be expected to work as is elsewhere.

I believe there was indeed a misunderstanding, especially as of commit f7418ad, this feature was present in the `WorkspaceSelector`, with the addition of functions that would connect and disconnect the observers. Then, the observer's status could be managed by the hosting GUI, in this case the `WorkspaceCalculator`. There, the logic was to disconnect them after a specified idle time, and in other cases when the interaction with `WorkspaceSelector` could be assumed to be finished. In this case, I am going to revert back to commit f7418ad and we can discuss the logic of deciding when the observers are disconnected.

**Comment by gvardany** [comment]
That idle time is exactly what I'm worried about, many things can happen during that. Ideally, we should disconnect immediately, once a choice is made in the selector. Clicking on it again can query the ADS again.

**Comment by gvardany** [comment]
I removed the disconnect on timeout, and cabled it to `activated` signal instead. Also made the event handlers in `WorkspaceSelector` synchronized. This is ready for another review.

File: qt/widgets/common/src/WorkspaceSelector.cpp [hunk 1]
```diff
@@ -125,6 +140,8 @@ void WorkspaceSelector::setSorted(bool sorted) {
   }
 }
 
+bool WorkspaceSelector::isConnected() { return m_connected; }
```
Comment (robertapplin) [inline qt/widgets/common/src/WorkspaceSelector.cpp:143]: can this method be `const`?

File: docs/source/release/v6.2.0/mantidworkbench.rst [hunk 2]
```diff
@@ -24,6 +24,7 @@ New and Improved
 - Script editor tab completion and call tip support for Numpy 1.21
 - The visibility of a component parameter in the Pick tab of the InstrumentViewer is now steered by the 'visible' atrribute of a parameter in IPF
 - Plot legends can be shown or hidden from the plot context menu.
+- ADS signal handlers are made synchronized in `WorkspaceSelector`. This reduces the probability of hard crash when interacting with the widget while a script manipulating a large number of workspaces is being run.
```
Comment (robertapplin) [inline docs/source/release/v6.2.0/mantidworkbench.rst:27]: I think this should be 'ADS signal handlers are now synchronized in the `WorkspaceSelector`'

**Review by robertapplin — CHANGES_REQUESTED** [review]
I can confirm that there is no crash for cases 1, 2 and 3. I also couldn't cause a crash for cases 4 and 5 - are we still looking for solutions to these? If so, is this PR still ready to go in without solutions to cases 4 and 5?

The code changes also look good, I just had one very minor suggestion

**Comment by gvardany** [comment]
Thanks for taking a look at this. Yes, I think this should not wait for the 4 and 5. Those are rare scenarios that can be investigated another time. Are you on release or debug build? Debug build doesn't crash as it is slow, so the race condition doesn't occur. It might also depend on the machine; if it is not so powerful, again the creation of workspace might take longer than the widget's reaction.

**Comment by robertapplin** [comment]
> Thanks for taking a look at this. Yes, I think this should not wait for the 4 and 5. Those are rare scenarios that can be investigated another time. Are you on release or debug build? Debug build doesn't crash as it is slow, so the race condition doesn't occur. It might also depend on the machine; if it is not so powerful, again the creation of workspace might take longer than the widget's reaction.

Ah ok, I was in debug mode but just tried with release and I can see the crash for case 4 and 5. But there is no crash for 1, 2, 3 👍

**Review by robertapplin — APPROVED** [review]
This is a difficult issue to fix but I think this is a good solution. I am also happy for edge cases 4 and 5 to be fixed at a later date

**Comment by martyngigg** [comment]
Do we have an issue recorded for looking into the edge cases?

**Review by martyngigg — APPROVED** [review]

**Comment by martyngigg** [comment]
One thing that could help track down 4/5 atleast as a starting point could be a `RelWithDebInfo` build and then in a shell before you start workbench activate core dump files:
```sh
build_dir> ulimit -c unlimited
build_dir> PYTHONPATH=$PWD/bin python3 bin/workbench
```
Make the crash happen and when it does it should produce a file called `core` in the `build_dir`. Open this with gdb:
```sh
build_dir> gdb /usr/bin/python3 core
```
and it should give the location of the segfault. This avoids needing to have the debugger running with the program which can slow it down and mean the bug is not produced.

**Comment by gvardany** [comment]
Just opened the follow-up issue https://github.com/mantidproject/mantid/issues/32308.

**Pull request merged**